internal/internal_workflow_client.go

```diff
@@ -1617,27 +1605,35 @@ func (w *workflowClientInterceptor) ExecuteWorkflow(
 		startRequest.WorkflowStartDelay = durationpb.New(in.Options.StartDelay)
 	}
 
-	var response *workflowservice.StartWorkflowExecutionResponse
-
 	grpcCtx, cancel := newGRPCContext(ctx, grpcMetricsHandler(
 		w.client.metricsHandler.WithTags(metrics.RPCTags(in.WorkflowType, metrics.NoneTagValue, in.Options.TaskQueue))),
 		defaultGrpcRetryParameters(ctx))
 	defer cancel()
 
-	response, err = w.client.workflowService.StartWorkflowExecution(grpcCtx, startRequest)
-	eagerWorkflowTask := response.GetEagerWorkflowTask()
-	if eagerWorkflowTask != nil && eagerExecutor != nil {
-		eagerExecutor.handleResponse(eagerWorkflowTask)
-	} else if eagerExecutor != nil {
-		eagerExecutor.release()
-	}
-	// Allow already-started error
 	var runID string
-	if e, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); ok && !in.Options.WorkflowExecutionErrorWhenAlreadyStarted {
-		runID = e.RunId
-	} else if err != nil {
-		return nil, err
+	if in.Options.WorkflowOperation == nil {
+		response, err := w.client.workflowService.StartWorkflowExecution(grpcCtx, startRequest)
+
+		eagerWorkflowTask := response.GetEagerWorkflowTask()
+		if eagerWorkflowTask != nil && eagerExecutor != nil {
+			eagerExecutor.handleResponse(eagerWorkflowTask)
+		} else if eagerExecutor != nil {
+			eagerExecutor.release()
+		}
+
+		// Allow already-started error
+		if e, ok := err.(*serviceerror.WorkflowExecutionAlreadyStarted); ok && !in.Options.WorkflowExecutionErrorWhenAlreadyStarted {
+			runID = e.RunId
+		} else if err != nil {
+			return nil, err
+		} else {
+			runID = response.RunId
+		}
 	} else {
+		response, err := w.executeWorkflowWithOperation(grpcCtx, startRequest, in.Options.WorkflowOperation)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
With regards to user-initiated retries, there is a catch here. Since the `startRequest` is re-created for each `ExecuteWorkflow` call - unless the user sets their own RequestID - the generated RequestID will be different each time. So in order to make the retry work, the user would need to use the UseExisting workflow id conflict policy. This will need to be documented.